pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/RealtimeOffsetAutoResetKafkaHandler.java

```diff
@@ -0,0 +1,135 @@
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Override this method to trigger Kafka Ecosystem APIs and replicate skipped offsets to the new topic.
+   * Then refer to the lagged topic's StreamConfig and return the new topic's stream config map.
+   */
+  protected abstract Map<String, String> triggerDataReplicationAndGetTopicInfo(
+      String tableNameWithType, StreamConfig streamConfig, String topicName, int partitionId, long fromOffset,
+      long toOffset);
+
+  public abstract void ensureBackfillJobsRunning(String tableNameWithType, List<String> topicNames);
+
+  public Collection<String> cleanupCompletedBackfillJobs(String tableNameWithType, Collection<String> topicNames) {
+    Collection<String> cleanedUpTopics = new ArrayList<>();
+    for (String topicName : topicNames) {
+      if (isTopicBackfillJobComplete(tableNameWithType, topicName)) {
+        cleanedUpTopics.add(topicName);
+      }
+    }
+    TableConfig currentTableConfig = _pinotHelixResourceManager.getTableConfig(tableNameWithType);
+    for (String topicName : cleanedUpTopics) {
+      removeTopicFromTableConfig(tableNameWithType, topicName, currentTableConfig);
+    }
+    try {
+      _pinotHelixResourceManager.setExistingTableConfig(currentTableConfig);
+    } catch (IOException e) {
+      LOGGER.error("Cannot remove backfill topics {} from the table config", topicNames, e);
+      cleanedUpTopics.clear();
+    }
+    return cleanedUpTopics;
+  }
+
+  public abstract boolean isTopicBackfillJobComplete(String tableNameWithType, String topicName);
+
+  private void addNewTopicToTableConfig(Map<String, String> streamConfig, TableConfig tableConfig) {
+    List<Map<String, String>> streamConfigs = IngestionConfigUtils.getStreamConfigMaps(tableConfig);
+    String topicNameKey = StreamConfigProperties.constructStreamProperty(
+        STREAM_TYPE, StreamConfigProperties.STREAM_TOPIC_NAME);
+    String topicName = streamConfig.get(topicNameKey);
+    Preconditions.checkNotNull(topicName);
+    for (Map<String, String> config : streamConfigs) {
+      if (topicName.equals(config.get(topicNameKey))) {
+        LOGGER.info("Topic {} already added to table {}", topicName, tableConfig.getTableName());
+        return;
+      }
+    }
+    streamConfig.put(StreamConfigProperties.EPHEMERAL_BACKFILL_TOPIC, String.valueOf(true));
+    streamConfig.put(
+        StreamConfigProperties.constructStreamProperty(
+            STREAM_TYPE, StreamConfigProperties.STREAM_CONSUMER_OFFSET_CRITERIA),
+        OffsetCriteria.SMALLEST_OFFSET_CRITERIA.getOffsetString());
+    IngestionConfigUtils.getStreamConfigMaps(tableConfig).add(streamConfig);
+  }
+
+  private void removeTopicFromTableConfig(String tableNameWithType, String topicName, TableConfig tableConfig) {
+    List<Map<String, String>> streamConfigMaps = IngestionConfigUtils.getStreamConfigMaps(tableConfig);
+    for (int i = 0; i < streamConfigMaps.size(); i++) {
```

Comment:
Major: 
Q1: Does streamConfigMaps maintain the order?
Q2: I feel it might be a concurrent modification exception here.